text/0000-box-and-in-for-stdlib.md

```diff
@@ -0,0 +1,520 @@
+    to write `box () (<expr>)` (or `box (alloc::HEAP) (<expr>)`) when
+    one wants to surround `<expr>` with parentheses.  It allows the
+    parser to provide clearer error messages if a user accidentally
+    writes `in <place-expr> <expr>`.
+
+ *  It would be premature for Rust to commit to any particular
+    protocol for supporting placement-`in`. A number of participants in
+    the discussion of [Placement Box RFC PR 470] were unhappy with the
+    baroque protocol, especially since it did not support DST and
+    potential future language changes would allow the protocol
+    proposed there to be significantly simplified.
+
+Therefore, this RFC proposes a middle ground for 1.0: Support the
+desired syntax, but do not provide stable support for end-user
+implementations of the operators. The only stable ways to use the
+overloaded `box <expr>` or `in (<place-expr>) <expr>` operators will be in
+tandem with types provided by the stdlib, such as `Box<T>`.
+
+# Detailed design
+
+* Add traits to `core::ops` for supporting the new operators.
+  This RFC does not commit to any particular set of traits,
+  since they are not currently meant to be implemented outside
+  of the stdlib. (However, a demonstration of one working set
+  of traits is given in [Appendix A].)
+
+  Any protocol that we adopt for the operators needs to properly
+  handle panics; i.e., `box <expr>` must properly cleanup any
+  intermediate state if `<expr>` panics during its evaluation,
+  and likewise for `in (<place-expr>) <expr>`
+
+  (See [Placement Box RFC PR 470] or [Appendix A] for discussion on
+   ways to accomplish this.)
+
+* Change `box <expr>` from built-in syntax (tightly integrated with
+  `Box<T>`) into an overloaded-`box` operator that uses the expected
+  return type to decide what kind of value to create.  For example, if
+  `Rc<T>` is extended with an implementation of the appropriate
+  operator trait, then
+
+  ```rust
+  let x: Rc<_> = box format!("Hello");
+  ```
+
+  could be a legal way to create an `Rc<String>` without having to
+  invoke the `Rc::new` function. This will be more efficient for
+  building instances of `Rc<T>` when `T` is a large type.  (It is also
+  arguably much cleaner syntax to read, regardless of the type `T`.)
+
+  Note that this change will require end-user code to no longer assume
+  that `box <expr>` always produces a `Box<T>`; such code will need to
+  either add a type annotation e.g. saying `Box<_>`, or will need to
+  call `Box::new(<expr>)` instead of using `box <expr>`.
+
+* Add support for parsing `in (<place-expr>) <expr>` as the basis for the
+  placement operator.
+
+  Remove support for `box (<place-expr>) <expr>` from the parser.
+
+  Make `in (<place-expr>) <expr>` an overloaded operator that uses
+  the `<place-expr>` to determine what placement code to run.
+
+* The only stablized implementation for the `box <expr>` operator
+  proposed by this RFC is `Box<T>`. The question of which other types
```

Comment:
I think it will too, I just haven't kept up with all the coherence rules and trait selection recently! FWIW this does not compile:

```rust
pub trait Foo {
    fn make() -> Self;
}

impl Foo for i32 {
    fn make() -> i32 { 2 }
}

fn main() {
    drop(Foo::make());
}
```

```
foo.rs:10:5: 10:9 error: unable to infer enough type information about `_`; type annotations required [E0282]
foo.rs:10     drop(Foo::make());
              ^~~~
error: aborting due to previous error
```